### PR TITLE
feat(affiliates): add composable dev/dapp mining apps for future automation

### LINKED ISSUE #2519: Implement some of the proposals for payout scripts to automate dev mining
Created by David Adams via monday.com integration. 🎉

### PR DESCRIPTION
**Pull request by daywiss**
**Motivation**

#2519

**Summary**

This adds a bunch of new "apps" which can be composed to create an automation pipeline. They can take input as a combination from standard in as json strings, env vars and a config file. These can be piped together using unix pipes to compose input and output. The idea is that they can be run independently if needed to verify parts of the pipeline. 

**Details**

For review purposes, you can ignore changes in the `scripts` dir, as these were just prototypes to get to this functionality. Production code changes are in `apps`, `libs/affiliates` and `tests/mocha`.

An example of how to run a dev mining pipeline:
` node apps/GetSheetsDevminingStatus.js | node apps/GenerateDevMiningConfig.js |  node apps/OpenDevMiningIssue.js | node apps/DevMiningRewards.js --network-mainnet_mnemonic  | node apps/DevMiningSaveResult`

This essentially pulls a whitelist from the google sheet, passes it into a config generator to generate proper config params to run the rewards. A github issue is opened with this information using a template, then params are passed to run the actual dev mining rewards. The output of this gets passed to a file saver, which formats the file as well as the filename and saves it to disk. 

If you could pick apart this pipeline, as a common use case would just be to verify dev mining. Assuming you had a config file you can run it like this:
`cat config.json | node apps/DevMiningRewards.js --network=mainnet_mnemonic` 

which will just output the results to console for you to compare.

If you just want to see what is currently whitelisted you can just run:

`node apps/GetSheetsDevminingStatus.js` (requires .env file)

Similarly you can compose a Dapp mining pipeline this way, though it is slightly different. 

The main point of this pattern is that each successive app only adds to an object that gets continually passed through to the next. The next app knows it can pull fields off an object, and add its own output parameters to it. Theres also an assumption that each app should add a unique key, and for now that requires some subjectivity.

Theres much more that needs to be done, including comments and more extensive tests, but most of this has been tested in production.  I will be working with @evaldofelipe to start automating this work. 

This also depends on #2501 going in.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [x]  Ran end-to-end test, running the code as in production
- [x]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

**Issue(s)**

Fixes #2519

### CODE REVIEW

File: packages/affiliates/apps/DappMiningRewards.js [hunk 1]
```diff
@@ -18,15 +20,12 @@ async function App(config) {
...
+    result
   };
-}
+};
 
-const config = Config();
-
-App(config)
-  .then(x => console.log(JSON.stringify(x, null, 2)))
+makeUnixPipe(App)
```
Comment (daywiss) [inline packages/affiliates/apps/DappMiningRewards.js:28]: this got converted to take its config from standard in. it expects a json object which is not documented yet.

File: packages/affiliates/apps/DevMiningRewards.js [hunk 2]
```diff
@@ -81,16 +85,13 @@ async function App(config, env) {
...
+    result
   };
-}
-
-const config = Config();
+};
 
-App(config, process.env)
-  .then(x => console.log(JSON.stringify(x, null, 2)))
+makeUnixPipe(App(process.env))
```
Comment (daywiss) [inline packages/affiliates/apps/DevMiningRewards.js:94]: again converted from a config file to a json blob piped through standard in

File: packages/affiliates/apps/MultiDappMiningRewards.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
+const lodash = require("lodash");
```
Comment (daywiss) [inline packages/affiliates/apps/MultiDappMiningRewards.js:1]: example of a pipeline that is all built into a single script. it is kind of nice but theres no way to call each stage in the pipeline independently without editing code, which is really annoying if you just want to generate the config file, or verify your github issue template looks correct.

File: packages/affiliates/apps/OpenDappMiningIssue.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+const App = env => async params => {
+  const { config } = params;
+  assert(config, "requires config");
+  const issueTemplate = await dappMiningTemplate(config);
+  const githubIssue = await createGithubIssue({ auth: env.github, ...issueTemplate });
+  return {
+    // data.number is the issue number
+    issueNumber: githubIssue.data.number,
+    issueTemplate,
+    ...params
```
Comment (daywiss) [inline packages/affiliates/apps/OpenDappMiningIssue.js:14]: this opens an issue, and also logs out the issue number so it can be consumed by another app. this would typically be needed when creating the PR to reference the issue number.

File: packages/affiliates/libs/affiliates/utils.js [hunk 5]
```diff
@@ -0,0 +1,281 @@
+// collection of functions to compose a automation pipeline
```
Comment (daywiss) [inline packages/affiliates/libs/affiliates/utils.js:1]: this utility file is just a bunch of helper functions that can be mostly unit tested

File: packages/affiliates/test/mocha/affiliates-utils.js [hunk 6]
```diff
@@ -0,0 +1,115 @@
+const { assert } = require("chai");
```
Comment (daywiss) [inline packages/affiliates/test/mocha/affiliates-utils.js:1]: unit tests for the utility file which needs to be expanded on, though most of the functions are tested in production

File: packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js [hunk 7]
```diff
@@ -1,5 +1,6 @@
 const { PriceFeedInterface } = require("./PriceFeedInterface");
 const { parseFixed } = require("@ethersproject/bignumber");
+const assert = require("assert");
```
Comment (daywiss) [inline packages/financial-templates-lib/src/price-feed/BasketSpreadPriceFeed.js:3]: everything below here can be ignored as they are part of #2501

**Comment by mrice32** [comment]
@daywiss is work still ongoing on this?

**Comment by daywiss** [comment]
@mrice32 yes trying to get this out of draft this week

File: packages/affiliates/apps/MultiDappMiningRewards.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+const App = env => async params => {
+  const web3 = getWeb3();
+
+  const empAbi = getAbi("ExpiringMultiParty");
+  const client = new BigQuery();
+  const queries = Queries({ client });
+
+  const dappmining = DappMining({ empAbi, queries, web3 });
+  params = lodash.castArray(params);
+  await Promise.map(params, async param => {
```
Comment (mrice32) [inline packages/affiliates/apps/MultiDappMiningRewards.js:31]: Are you expecting param to be a specific thing? I really like the idea of piping between these different "apps", but I think it would be nice to document what each expects as an input and what it outputs since unlike most unix CLI tools, this looks like it's expecting something a bit more specified than raw text.

As another side note: I think variable names like `param` in general are somewhat opaque. Would be nice to have a more friendly name so it's a little more clear what type of param is expected since I'm assuming there are some requirements.

File: packages/affiliates/libs/affiliates/utils.js [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+//   payoutAddress,
+//   name,
+//   identifier,
+//   enabled
+// }]
+// And returns it in the whitelist format consumable by dev mining:
+// [
+//   [ address, payoutAddress],
+//   [ address, payoutAddress],
+// ]
```
Comment (mrice32) [inline packages/affiliates/libs/affiliates/utils.js:164]: +1 Until we move everything to ts, this is super helpful!

File: packages/affiliates/libs/affiliates/utils.js [hunk 10]
```diff
@@ -0,0 +1,323 @@
+// collection of functions to compose a automation pipeline with regards to dev/dapp mining.
```
Comment (mrice32) [inline packages/affiliates/libs/affiliates/utils.js:1]: High-level comment: these function feel more contextual than just utils. Not saying we should change the structure, but I could imagine these being split out into more well-defined libraries or class-type objects that have a more refined purpose.

For instance, the config generation functions seem quite specialized and maybe more suited for a special file or class just dedicated to generating configs.

**Review by mrice32 — APPROVED** [review]
Looks great! Only minor comments. Really like the idea of using piping this way!!

File: packages/affiliates/apps/MultiDappMiningRewards.js (hunk 8 above)
Comment (nicholaspai) [reply packages/affiliates/apps/MultiDappMiningRewards.js:31]: Agreed with matt on opaque naming. Cool use of `Promise.map` I'm going to start replacing my Promise.all's with this.

Comment (daywiss) [reply packages/affiliates/apps/MultiDappMiningRewards.js:31]: yes this is my bad, params is a bad name and needs to be documented. I have some thoughts around making this more of a common convention to send data between each app to make it more clear, and also be able to throw meaningful errors when expected data is missing. Also yes typescript would be really useful here to both document and enforce these types.  

I will update this with better comments and naming.

File: packages/affiliates/libs/affiliates/utils.js (hunk 10 above)
Comment (daywiss) [reply packages/affiliates/libs/affiliates/utils.js:1]: I agree, though I dont think i want to do this for this PR.  This could be organized better ( though im not sure how yet), i'm using utils as a catch all for these loosely affiliated static functions.

File: packages/affiliates/apps/DevMiningRewards.js (hunk 2 above)
Comment (chrismaree) [reply packages/affiliates/apps/DevMiningRewards.js:94]: this is cool.

File: packages/affiliates/payouts/2021-02-08_2021-02-15.json [hunk 11]
```diff
@@ -1,80 +0,0 @@
-{
```
Comment (chrismaree) [inline packages/affiliates/payouts/2021-02-08_2021-02-15.json:1]: why was this file removed?

**Review by chrismaree — APPROVED** [review]
This is awesome! great work on these scripts @daywiss . I cant wait until these processes are all automated. I really like your piping technique that you've got going here.

File: packages/affiliates/libs/affiliates/utils.js (hunk 10 above)
Comment (mrice32) [reply packages/affiliates/libs/affiliates/utils.js:1]: SG!

File: packages/affiliates/payouts/2021-02-08_2021-02-15.json (hunk 11 above)
Comment (daywiss) [reply packages/affiliates/payouts/2021-02-08_2021-02-15.json:1]: good question. i think this was added accidentally in a past PR!

Comment (daywiss) [reply packages/affiliates/payouts/2021-02-08_2021-02-15.json:1]: actually looking at it now, i think it was committed to the wrong directory, because its not in payouts/devmining where it should be. thanks for noticing this. will move it rather than delete
